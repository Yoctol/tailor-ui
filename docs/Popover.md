---
id: popover
title: Popover
---

The floating card popped by clicking or hovering.

## When To Use

A simple popup menu to provide extra information or operations.

Comparing with `Tooltip`, besides information `Popover` card can also provide action elements like links and buttons.

## Examples

```js
import { Popover } from 'tailor-ui';
```

### Basic

```jsx live
<Popover title="Popover Title" content="Popover Content">
  <Button>Button</Button>
</Popover>
```

### Positioning

```jsx live
<Box px="5">
  <Flex justifyContent="space-between">
    <Popover
      position={Position.TOP_LEFT}
      title="Top Left"
      content={
        <Flex
          alignItems="center"
          justifyContent="center"
          width="200px"
          height="200px"
        >
          TOP_LEFT
        </Flex>
      }
    >
      <Button>TOP_LEFT</Button>
    </Popover>

    <Popover
      position={Position.TOP}
      title="Top"
      content={
        <Flex
          alignItems="center"
          justifyContent="center"
          width="200px"
          height="200px"
        >
          TOP
        </Flex>
      }
    >
      <Button>TOP</Button>
    </Popover>

    <Popover
      position={Position.TOP_RIGHT}
      title="Top Right"
      content={
        <Flex
          alignItems="center"
          justifyContent="center"
          width="200px"
          height="200px"
        >
          TOP_RIGHT
        </Flex>
      }
    >
      <Button>TOP_RIGHT</Button>
    </Popover>
  </Flex>

  <Flex justifyContent="space-between" mt="5">
    <Popover
      position={Position.LEFT}
      title="Left"
      content={
        <Flex
          alignItems="center"
          justifyContent="center"
          width="200px"
          height="200px"
        >
          LEFT
        </Flex>
      }
    >
      <Button>LEFT</Button>
    </Popover>

    <Popover
      position={Position.RIGHT}
      title="Right"
      content={
        <Flex
          alignItems="center"
          justifyContent="center"
          width="200px"
          height="200px"
        >
          RIGHT
        </Flex>
      }
    >
      <Button>RIGHT</Button>
    </Popover>
  </Flex>

  <Flex justifyContent="space-between" mt="5">
    <Popover
      position={Position.BOTTOM_LEFT}
      title="Bottom Left"
      content={
        <Flex
          alignItems="center"
          justifyContent="center"
          width="200px"
          height="200px"
        >
          BOTTOM_LEFT
        </Flex>
      }
    >
      <Button>BOTTOM_LEFT</Button>
    </Popover>

    <Popover
      position={Position.BOTTOM}
      title="Bottom"
      content={
        <Flex
          alignItems="center"
          justifyContent="center"
          width="200px"
          height="200px"
        >
          BOTTOM
        </Flex>
      }
    >
      <Button>BOTTOM</Button>
    </Popover>

    <Popover
      position={Position.BOTTOM_RIGHT}
      title="Bottom Right"
      content={
        <Flex
          alignItems="center"
          justifyContent="center"
          width="200px"
          height="200px"
        >
          BOTTOM_RIGHT
        </Flex>
      }
    >
      <Button>BOTTOM_RIGHT</Button>
    </Popover>
  </Flex>
</Box>
```

### With content hide

```jsx live
<Popover
  position={Position.RIGHT}
  title="Title"
  content={hide => (
    <div>
      Popover Content
      <br />
      <Button size="sm" onClick={hide}>
        Hide Popover
      </Button>
    </div>
  )}
>
  <Button>Button</Button>
</Popover>
```

### With onVisibleChange

```jsx live
<Popover
  position={Position.RIGHT}
  title="Title"
  content="Popover content"
  onVisibleChange={console.log}
>
  <Button>Button</Button>
</Popover>
```

### Popover with Select

```jsx live
() => {
  const [value, setValue] = useState('Banana');

  return (
    <Popover
      position={Position.RIGHT}
      content={
        <Select
          value={value}
          onChange={setValue}
          options={['Banana', 'Orange', 'Apple', 'Mango']}
        />
      }
    >
      <Button>Button</Button>
    </Popover>
  );
}
```

### Nested Usage

```jsx live
<Popover
  position={Position.TOP}
  title="Nest 1"
  content={
    <Flex
      alignItems="center"
      justifyContent="center"
      width="200px"
      height="200px"
    >
      <Popover
        position={Position.RIGHT}
        title="Nest 2"
        content={
          <Flex
            alignItems="center"
            justifyContent="center"
            width="200px"
            height="200px"
          >
            <Popover
              position={Position.BOTTOM}
              title="Nest 3"
              content={
                <Flex
                  alignItems="center"
                  justifyContent="center"
                  width="200px"
                  height="200px"
                >
                  <Popover
                    position={Position.LEFT}
                    title="Nest 4"
                    content={
                      <Flex
                        alignItems="center"
                        justifyContent="center"
                        width="200px"
                        height="200px"
                      >
                        Last
                      </Flex>
                    }
                  >
                    <Button>Button</Button>
                  </Popover>
                </Flex>
              }
            >
              <Button>Button</Button>
            </Popover>
          </Flex>
        }
      >
        <Button>Button</Button>
      </Popover>
    </Flex>
  }
>
  <Button>Button</Button>
</Popover>
```

## API
