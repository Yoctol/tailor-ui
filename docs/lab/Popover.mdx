---
name: Popover
menu: Lab
route: /lab/popover
---

import { PropsTable, Playground, Link } from 'docz';
import { Button, Position } from 'tailor-ui';
import { Popover } from '@tailor-ui/lab';

# Popover

The floating card popped by clicking or hovering.

## When To Use

A simple popup menu to provide extra information or operations.

Comparing with `Tooltip`, besides information `Popover` card can also provide action elements like links and buttons.

## Examples

```js
import { Popover } from 'tailor-ui';
```

### Basic

<Playground>
  <Popover title="Popover Title" content="Popover Content">
    <Button>Button</Button>
  </Popover>
</Playground>

### With position right

<Playground>
  <Popover
    position={Position.RIGHT}
    title="Popover Title"
    content="Popover Content"
  >
    <Button>Button</Button>
  </Popover>
</Playground>

### With content hide

<Playground>
  <Popover
    position={Position.RIGHT}
    title="Title"
    content={hide => (
      <div>
        Popover Content
        <br />
        <Button size="sm" onClick={hide}>
          Hide Popover
        </Button>
      </div>
    )}
  >
    <Button>Button</Button>
  </Popover>
</Playground>

### With onVisibleChange

<Playground>
  <Popover
    position={Position.RIGHT}
    title="Title"
    content="Popover content"
    onVisibleChange={console.log}
  >
    <Button>Button</Button>
  </Popover>
</Playground>

## API

<PropsTable of={Popover} />
