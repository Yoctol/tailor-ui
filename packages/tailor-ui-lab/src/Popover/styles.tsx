import styled from 'styled-components';
import {
  MinHeightProps,
  MinWidthProps,
  SpaceProps,
  minHeight,
  minWidth,
  space,
} from 'styled-system';

export type StyledPopoverProps = SpaceProps & MinWidthProps & MinHeightProps;

export const StyledPopover = styled.div<StyledPopoverProps>`
  padding: ${p => p.theme.space[2]} ${p => p.theme.space[3]};
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.lg};
  border-color: ${p => p.theme.colors.gray300};
  background-color: ${p => p.theme.colors.light};
  box-shadow: 0 2px 4px rgba(100, 120, 168, 0.3);
  color: ${p => p.theme.colors.gray700};
  font-size: ${p => p.theme.fontSizes.sm};
  text-align: left;
  white-space: nowrap;

  ${space}
  ${minWidth}
  ${minHeight}
`;

export const PopoverHeader = styled.div`
  margin-bottom: ${p => p.theme.space[2]};
`;
