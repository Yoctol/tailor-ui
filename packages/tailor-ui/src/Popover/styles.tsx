import styled from 'styled-components';
import {
  MinHeightProps,
  MinWidthProps,
  SpaceProps,
  minHeight,
  minWidth,
  space,
} from 'styled-system';

export type StyledPopoverProps = SpaceProps & MinWidthProps & MinHeightProps;

export const StyledPopover = styled.div<StyledPopoverProps>`
  padding: ${p => p.theme.space[2]} ${p => p.theme.space[3]};
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.lg};
  border-color: ${p => p.theme.colors.gray300};
  background-color: ${p => p.theme.colors.light};
  box-shadow: ${p => p.theme.shadows.base};
  color: ${p => p.theme.colors.gray700};
  font-size: ${p => p.theme.fontSizes.sm};
  text-align: left;
  white-space: nowrap;

  ${space}
  ${minWidth}
  ${minHeight}
`;

export const PopoverHeader = styled.div`
  margin-bottom: ${p => p.theme.space[2]};
`;
