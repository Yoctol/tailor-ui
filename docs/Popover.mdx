---
name: Popover
menu: Components
route: /components/popover
---

import { PropsTable, Playground, Link } from 'docz';
import { Button, Popover, Box, Flex, Position } from 'tailor-ui';
import { Select } from '@tailor-ui/lab';

# Popover

The floating card popped by clicking or hovering.

## When To Use

A simple popup menu to provide extra information or operations.

Comparing with `Tooltip`, besides information `Popover` card can also provide action elements like links and buttons.

## Examples

```js
import { Popover } from 'tailor-ui';
```

### Basic

<Playground>
  <Popover title="Popover Title" content="Popover Content">
    <Button>Button</Button>
  </Popover>
</Playground>

### Positioning

<Playground>
  <Box px="8">
    <Flex justifyContent="space-between">
      <Popover
        position={Position.TOP_LEFT}
        title="Top Left"
        content={
          <Flex alignItems="center" justifyContent="center" width="200px" height="200px">
            TOP_LEFT
          </Flex>
        }
      >
        <Button>TOP_LEFT</Button>
      </Popover>

      <Popover
        position={Position.TOP}
        title="Top"
        content={
          <Flex alignItems="center" justifyContent="center" width="200px" height="200px">
            TOP
          </Flex>
        }
      >
        <Button>TOP</Button>
      </Popover>

      <Popover
        position={Position.TOP_RIGHT}
        title="Top Right"
        content={
          <Flex alignItems="center" justifyContent="center" width="200px" height="200px">
            TOP_RIGHT
          </Flex>
        }
      >
        <Button>TOP_RIGHT</Button>
      </Popover>
    </Flex>

    <Flex justifyContent="space-between" mt="8">
      <Popover
        position={Position.LEFT}
        title="Left"
        content={
          <Flex alignItems="center" justifyContent="center" width="200px" height="200px">
            LEFT
          </Flex>
        }
      >
        <Button>LEFT</Button>
      </Popover>

      <Popover
        position={Position.RIGHT}
        title="Right"
        content={
          <Flex alignItems="center" justifyContent="center" width="200px" height="200px">
            RIGHT
          </Flex>
        }
      >
        <Button>RIGHT</Button>
      </Popover>
    </Flex>

    <Flex justifyContent="space-between" mt="8">
      <Popover
        position={Position.BOTTOM_LEFT}
        title="Bottom Left"
        content={
          <Flex alignItems="center" justifyContent="center" width="200px" height="200px">
            BOTTOM_LEFT
          </Flex>
        }
      >
        <Button>BOTTOM_LEFT</Button>
      </Popover>

      <Popover
        position={Position.BOTTOM}
        title="Bottom"
        content={
          <Flex alignItems="center" justifyContent="center" width="200px" height="200px">
            BOTTOM
          </Flex>
        }
      >
        <Button>BOTTOM</Button>
      </Popover>

      <Popover
        position={Position.BOTTOM_RIGHT}
        title="Bottom Right"
        content={
          <Flex alignItems="center" justifyContent="center" width="200px" height="200px">
            BOTTOM_RIGHT
          </Flex>
        }
      >
        <Button>BOTTOM_RIGHT</Button>
      </Popover>
    </Flex>

  </Box>
</Playground>

### With content hide

<Playground>
  <Popover
    position={Position.RIGHT}
    title="Title"
    content={hide => (
      <div>
        Popover Content
        <br />
        <Button size="sm" onClick={hide}>
          Hide Popover
        </Button>
      </div>
    )}
  >
    <Button>Button</Button>
  </Popover>
</Playground>

### With onVisibleChange

<Playground>
  <Popover
    position={Position.RIGHT}
    title="Title"
    content="Popover content"
    onVisibleChange={console.log}
  >
    <Button>Button</Button>
  </Popover>
</Playground>

### Popover with Select

<Playground>
  {() => {
    const [value, setValue] = React.useState('Banana');

    return (
      <Popover
        position={Position.RIGHT}
        content={
          <Select
            value={value}
            onChange={setValue}
            options={['Banana', 'Orange', 'Apple', 'Mango']}
          />
        }
      >
        <Button>Button</Button>
      </Popover>
    );

}}

</Playground>

### Nested Usage

<Playground>
  <Popover
    position={Position.TOP}
    title="Nest 1"
    content={
      <Flex
        alignItems="center"
        justifyContent="center"
        width="200px"
        height="200px"
      >
        <Popover
          position={Position.RIGHT}
          title="Nest 2"
          content={
            <Flex
              alignItems="center"
              justifyContent="center"
              width="200px"
              height="200px"
            >
              <Popover
                position={Position.BOTTOM}
                title="Nest 3"
                content={
                  <Flex
                    alignItems="center"
                    justifyContent="center"
                    width="200px"
                    height="200px"
                  >
                    <Popover
                      position={Position.LEFT}
                      title="Nest 4"
                      content={
                        <Flex
                          alignItems="center"
                          justifyContent="center"
                          width="200px"
                          height="200px"
                        >
                          Last
                        </Flex>
                      }
                    >
                      <Button>Button</Button>
                    </Popover>
                  </Flex>
                }
              >
                <Button>Button</Button>
              </Popover>
            </Flex>
          }
        >
          <Button>Button</Button>
        </Popover>
      </Flex>
    }
  >
    <Button>Button</Button>
  </Popover>
</Playground>

## API

<PropsTable of={Popover} />
