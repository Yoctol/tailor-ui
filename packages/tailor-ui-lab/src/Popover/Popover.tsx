import React, {
  CSSProperties,
  FunctionComponent,
  ReactNode,
  RefObject,
  cloneElement,
  forwardRef,
  isValidElement,
  memo,
  useRef,
  useState,
} from 'react';
import { animated } from 'react-spring';
import { mergeEventProps } from '@tailor-ui/utils';

import {
  Heading,
  Position,
  Positioner,
  Positions,
  useClickOutside,
  useKeydown,
} from 'tailor-ui';

import { PopoverContent, PopoverHeader, StyledPopover } from './styles';

interface PopoverPopup {
  style: CSSProperties;
  title: ReactNode | ((handleClose: () => void) => ReactNode);
  content: ReactNode | ((handleClose: () => void) => ReactNode);
  handleClose: () => void;
}

const PopoverPopup = memo(
  forwardRef<HTMLDivElement, PopoverPopup>(function PopoverPopup(
    { style, title, content, handleClose, ...otherProps },
    ref
  ) {
    return (
      <animated.div style={style} ref={ref}>
        <StyledPopover {...otherProps}>
          {title && (
            <PopoverHeader>
              <Heading.h6>
                {title instanceof Function ? title(handleClose) : title}
              </Heading.h6>
            </PopoverHeader>
          )}
          <PopoverContent>
            {content instanceof Function ? content(handleClose) : content}
          </PopoverContent>
        </StyledPopover>
      </animated.div>
    );
  })
);

export interface PopoverProps {
  /**
   * Whether the floating popover card is visible by default. Only support when the trigger is `click`
   */
  defaultVisible?: boolean;
  /**
   * Whether the floating popover card is visible. Only support when the trigger is `click`
   */
  visible?: boolean;
  /**
   * 	Callback executed when visibility of the popover card is changed
   */
  onVisibleChange?: (visible: boolean) => void;
  /**
   * The position base on the children component
   */
  position?: Positions;
  title?: ReactNode | ((handleClose: () => void) => ReactNode);
  /**
   * A string or react component inside this popover.
   * If you are using click to trigger, it can be a
   * function that with `hide` callback as first argument
   */
  content: ReactNode | ((handleClose: () => void) => ReactNode);
}

const Popover: FunctionComponent<PopoverProps> = ({
  children,
  position = Position.TOP,
  title,
  content,
  defaultVisible = false,
  visible: visibleFromProps,
  onVisibleChange,
}) => {
  const childrenRef = useRef(null);
  const popupRef = useRef(null);
  const [visibleFromSelf, setVisibleFromSelf] = useState(defaultVisible);

  const hasVisibleFromProps = typeof visibleFromProps !== 'undefined';

  const visible = hasVisibleFromProps ? visibleFromProps : visibleFromSelf;

  const handleOpen = () => {
    if (onVisibleChange) {
      onVisibleChange(true);
    }

    if (!hasVisibleFromProps) {
      setVisibleFromSelf(true);
    }
  };

  const handleClose = () => {
    if (onVisibleChange) {
      onVisibleChange(false);
    }

    if (!hasVisibleFromProps) {
      setVisibleFromSelf(false);
    }
  };

  const toggle = () => {
    if (visible) {
      handleClose();
    } else {
      handleOpen();
    }
  };

  useClickOutside({
    listening: visible,
    refs: [childrenRef, popupRef],
    onClickOutside: handleClose,
  });

  useKeydown({
    listening: visible,
    keyCode: 27,
    onKeydown: handleClose,
  });

  const renderChildren = ({ ref }: { ref: RefObject<HTMLElement> }) => {
    if (children instanceof Function) {
      return children({
        ref,
        bind: (props: any) =>
          mergeEventProps(props, {
            onClick: toggle,
          }),
      });
    }

    if (!isValidElement<any>(children)) {
      return children;
    }

    return cloneElement(children, {
      ref,
      ...mergeEventProps(children.props, {
        onClick: toggle,
      }),
    });
  };

  return (
    <Positioner
      positionerRef={popupRef}
      targetRef={childrenRef}
      visible={visible}
      position={position}
      positioner={({ style }) => (
        <PopoverPopup
          ref={popupRef}
          style={style}
          title={title}
          content={content}
          handleClose={handleClose}
        />
      )}
    >
      {renderChildren}
    </Positioner>
  );
};

export default Popover;
