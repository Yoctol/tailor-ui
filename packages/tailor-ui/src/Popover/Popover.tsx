import React, {
  CSSProperties,
  FunctionComponent,
  ReactNode,
  createContext,
  forwardRef,
  memo,
  useContext,
  useRef,
  useState,
} from 'react';
import { animated } from 'react-spring';

import {
  useClickOutside,
  useKeydown,
  useRenderChildren,
  useTargetRef,
  useToggleTrigger,
} from '@tailor-ui/hooks';

import Positioner from '../Positioner';
import { Heading } from '../Typography';
import { Position, Positions } from '../constants';

import { PopoverHeader, StyledPopover, StyledPopoverProps } from './styles';

interface PopoverPopup {
  style: CSSProperties;
  title: ReactNode | ((handleClose: () => void) => ReactNode);
  content: ReactNode | ((handleClose: () => void) => ReactNode);
  handleClose: () => void;
}

const PopoverPopup = memo(
  forwardRef<HTMLDivElement, PopoverPopup>(function PopoverPopup(
    { style, title, content, handleClose, ...otherProps },
    ref
  ) {
    return (
      <animated.div style={style} ref={ref}>
        <StyledPopover {...otherProps}>
          {title && (
            <PopoverHeader>
              <Heading.h6>
                {title instanceof Function ? title(handleClose) : title}
              </Heading.h6>
            </PopoverHeader>
          )}
          {content instanceof Function ? content(handleClose) : content}
        </StyledPopover>
      </animated.div>
    );
  })
);

export type PopoverProps = StyledPopoverProps & {
  /**
   * Whether the floating popover card is visible by default. Only support when the trigger is `click`
   */
  defaultVisible?: boolean;
  /**
   * Whether the floating popover card is visible. Only support when the trigger is `click`
   */
  visible?: boolean;
  /**
   * 	Callback executed when visibility of the popover card is changed
   */
  onVisibleChange?: (visible: boolean) => void;
  /**
   * The position base on the children component
   */
  position?: Positions;
  /**
   * A string or react component inside this popover.
   * If you are using click to trigger, it can be a
   * function that with `hide` callback as first argument
   */
  title?: ReactNode | ((handleClose: () => void) => ReactNode);
  /**
   * A string or react component inside this popover.
   * If you are using click to trigger, it can be a
   * function that with `hide` callback as first argument
   */
  content: ReactNode | ((handleClose: () => void) => ReactNode);
  onOpenComplete?: () => void;
  onCloseComplete?: () => void;
};

const ClickOutsideContext = createContext<{
  setHasChild: (hasChild: boolean) => void;
}>({
  setHasChild: () => {},
});

const Popover: FunctionComponent<PopoverProps> = ({
  children,
  position = Position.TOP,
  title,
  content,
  defaultVisible = false,
  visible: visibleFromProps,
  onVisibleChange,
  onOpenComplete,
  onCloseComplete,
  ...otherProps
}) => {
  const targetRef = useTargetRef({
    children,
  });
  const popupRef = useRef(null);

  const { setHasChild: setHasChildFromContext } = useContext(
    ClickOutsideContext
  );
  const [hasChild, setHasChild] = useState(false);

  const { visible, handleClose, toggle } = useToggleTrigger({
    visible: visibleFromProps,
    defaultVisible,
    onVisibleChange: newVisible => {
      setHasChildFromContext(newVisible);
      if (onVisibleChange) {
        onVisibleChange(newVisible);
      }
    },
  });

  useClickOutside({
    listening: hasChild ? false : visible,
    refs: [targetRef, popupRef],
    onClickOutside: handleClose,
  });

  useKeydown({
    listening: visible,
    keyCode: 27,
    onKeydown: handleClose,
  });

  const renderChildren = useRenderChildren({
    targetRef,
    children,
    mergeProps: {
      onClick: toggle,
    },
  });

  return (
    <ClickOutsideContext.Provider value={{ setHasChild }}>
      <Positioner
        positionerRef={popupRef}
        targetRef={targetRef}
        visible={visible}
        onOpenComplete={onOpenComplete}
        onCloseComplete={onCloseComplete}
        position={position}
        positioner={({ style }) => (
          <PopoverPopup
            ref={popupRef}
            style={style}
            title={title}
            content={content}
            handleClose={handleClose}
            {...otherProps}
          />
        )}
      >
        {renderChildren}
      </Positioner>
    </ClickOutsideContext.Provider>
  );
};

export default Popover;
