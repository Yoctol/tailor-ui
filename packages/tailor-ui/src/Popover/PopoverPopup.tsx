import React, { ComponentType, ReactNode, memo } from 'react';

import { Heading } from '../Typography';

import { PopoverHeader, StyledPopover } from './styles';

interface PopoverPopup {
  Wrapper?: ComponentType;
  title: ReactNode | ((handleClose: () => void) => ReactNode);
  content: ReactNode | ((handleClose: () => void) => ReactNode);
  handleClose: () => void;
}

const PopoverPopup = memo<PopoverPopup>(function PopoverPopup({
  title,
  content,
  handleClose,
  Wrapper = StyledPopover,
  ...otherProps
}) {
  return (
    <Wrapper {...otherProps}>
      {title && (
        <PopoverHeader>
          <Heading.h6>
            {title instanceof Function ? title(handleClose) : title}
          </Heading.h6>
        </PopoverHeader>
      )}
      {content instanceof Function ? content(handleClose) : content}
    </Wrapper>
  );
});

export default PopoverPopup;
