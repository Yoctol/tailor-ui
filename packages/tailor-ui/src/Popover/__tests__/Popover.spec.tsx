import React from 'react';

import { fireEvent, render, wait } from 'test/test-utils';

import { Popover } from '../Popover';

describe('Popover', () => {
  it('should render correctly', () => {
    const { baseElement } = render(
      <Popover visible title="Popover Title" content="Popover Content">
        <button type="button">button</button>
      </Popover>
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should render to document when click children', () => {
    const { getByText } = render(
      <Popover title="Popover Title" content="Popover Content">
        <button type="button">button</button>
      </Popover>
    );

    const button = getByText('button');

    fireEvent.click(button);

    const popconfirm = getByText('Popover Content');

    expect(popconfirm).toBeInTheDocument();
  });

  it('should not visible when click children twice', () => {
    const { getByText } = render(
      <Popover title="Popover Title" content="Popover Content">
        <button type="button">button</button>
      </Popover>
    );

    const button = getByText('button');

    fireEvent.click(button);
    fireEvent.click(button);

    const popconfirm = getByText('Popover Content');

    expect(popconfirm).not.toBeVisible();
  });

  it('should not visible when click hide button', () => {
    const { getByText } = render(
      <Popover
        title="Popover Title"
        content={hide => (
          <>
            Popover Content
            <button type="button" onClick={hide}>
              hide
            </button>
          </>
        )}
      >
        <button type="button">button</button>
      </Popover>
    );

    const button = getByText('button');
    fireEvent.click(button);

    const hideButton = getByText('hide');
    fireEvent.click(hideButton);

    const popconfirm = getByText('Popover Content');

    expect(popconfirm).not.toBeVisible();
  });

  it('should not visible when click outside', () => {
    const { getByText } = render(
      <div>
        <Popover title="Popover Title" content="Popover Content">
          <button type="button">button</button>
        </Popover>
        <button type="button">outside</button>
      </div>
    );

    const button = getByText('button');
    fireEvent.click(button);

    const outsideButton = getByText('outside');
    fireEvent.click(outsideButton);

    const popconfirm = getByText('Popover Content');

    expect(popconfirm).not.toBeVisible();
  });

  it('should close latest one when nested usage', async () => {
    const { getByText } = render(
      <div>
        <Popover
          title="popover1"
          content={
            <Popover
              title="popover2"
              content={
                <Popover title="popover3" content="latest">
                  <button type="button">button3</button>
                </Popover>
              }
            >
              <button type="button">button2</button>
            </Popover>
          }
        >
          <button type="button">button1</button>
        </Popover>
        <button type="button">outside</button>
      </div>
    );

    const button1 = getByText('button1');
    fireEvent.click(button1);
    const button2 = getByText('button2');
    fireEvent.click(button2);
    const button3 = getByText('button3');
    fireEvent.click(button3);

    const popover1 = getByText('popover1');
    const popover2 = getByText('popover2');
    const popover3 = getByText('popover3');
    await wait(() => [
      expect(popover1).toBeVisible(),
      expect(popover2).toBeVisible(),
      expect(popover3).toBeVisible(),
    ]);

    const outsideButton = getByText('outside');
    fireEvent.click(outsideButton);

    await wait(() => [
      expect(popover1).toBeVisible(),
      expect(popover2).toBeVisible(),
      expect(popover3).not.toBeVisible(),
    ]);
  });
});
