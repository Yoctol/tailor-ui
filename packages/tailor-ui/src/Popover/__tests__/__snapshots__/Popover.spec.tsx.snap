// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render correctly 1`] = `
.c3 {
  font-size: 0.875rem;
  margin: 0;
  color: #373737;
  font-weight: bold;
  line-height: 1.5;
}

.c0 {
  position: absolute;
  top: 0;
  left: 0;
  will-change: transform;
}

.c1 {
  padding: 0.5rem 1rem;
  border: 1px solid;
  border-radius: 0.5rem;
  border-color: #e0e0e0;
  background-color: #ffffff;
  box-shadow: 0 2px 4px rgba(100,120,168,0.3);
  color: #373737;
  font-size: 0.875rem;
  text-align: left;
  white-space: nowrap;
}

.c2 {
  margin-bottom: 0.5rem;
}

<body>
  <div>
    <button
      type="button"
    >
      button
    </button>
  </div>
  <div
    portal-container=""
  >
    <div
      style="position: fixed; top: 0px; left: 0px; z-index: 10;"
    >
      <div
        class="c0"
        style="transform: translate3d(-6px, -6px, 0px);"
      >
        <div
          style="opacity: 0; transform: scale(0.9);"
        >
          <div
            class="c1"
          >
            <div
              class="c2"
            >
              <h6
                class="c3"
              >
                Popover Title
              </h6>
            </div>
            Popover Content
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
